projects/components/src/dropdown/dynamic-dropdown-position.directive.ts

```diff
@@ -0,0 +1,173 @@
+/*!
+ * Copyright 2020 VMware, Inc.
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+import { DOCUMENT } from '@angular/common';
+import {
+    AfterContentInit,
+    ContentChild,
+    Directive,
+    ElementRef,
+    HostListener,
+    Inject,
+    OnDestroy,
+    Renderer2,
+} from '@angular/core';
+import { ClrDropdown, ClrDropdownMenu } from '@clr/angular';
+
+const CONTENT_AREA_SELECTOR = '.content-area';
+const NO_SCROLLING_CLASSNAME = 'no-scrolling'; // Set by Clarity when a modal is opened
+
+/**
+ * Directive for ClrDropDown which repositions the drop-down menu dynamically based on the available space.
+ *
+ * If there is not enough space to show the whole drop-down menu, then leave the default position which is `bottom` and
+ * users have to scroll down to access it.
+ *
+ * @usageNotes
+ *
+ * ```
+ * <clr-dropdown vcdDropdown>
+ *     <button class="btn btn-link vm-power-link" clrDropdownTrigger>
+ *         {{ 'vm.actions' | translate }}
+ *         <clr-icon shape="caret down"></clr-icon>
+ *         </button>
+ *     <clr-dropdown-menu *clrIfOpen>
+ *         <label class="dropdown-header">Dropdown header</label>
+ *         <button type="button" clrDropdownItem>Action 1</button>
+ *         <button type="button" clrDropdownItem>Action 2</button>
+ *         <div class="dropdown-divider"></div>
+ *         <button type="button" clrDropdownItem>Link 1</button>
+ *         <button type="button" clrDropdownItem>Link 2</button>
+ *     </clr-dropdown-menu>
+ * </clr-dropdown>
+ * ```
+ */
+/**
+ * This component is marked as dynamic to suppress the '"strictMetadataEmit": true' error being reported on DI of Document
+ */
+/** @dynamic */
+@Directive({
+    selector: 'clr-dropdown[vcdDynamicDropdown]',
+})
+export class DynamicDropdownPositionDirective implements AfterContentInit, OnDestroy {
+    private contentAreaElement: HTMLElement;
+    private dropdownTriggerElement;
+    private dropdownMenuElement;
+    // Drop downs should be position absolutely to the window when they are in modal
+    private isInsideModal = false;
+
+    @ContentChild(ClrDropdownMenu, { static: false })
+    set dropdownMenu(dropdown: ClrDropdownMenu) {
+        if (dropdown) {
+            this.dropdownTriggerElement = this.elRef.nativeElement;
+            this.dropdownMenuElement = this.dropdownTriggerElement.querySelector('clr-dropdown-menu');
+            if (this.dropdownMenuElement) {
+                // Recalculate the dropdown position on open
+                this.resetPosition(this.dropdownMenuElement, this.positionTop, this.positionLeft);
+            }
+            try {
+                // Internal API
+                (dropdown as any).popoverInstance.removeScrollEventListeners();
+            } catch (e) {
+                console.warn('Clarity has changed ClrDropdownMenu internal API. Please fix me!');
```

Comment:
Should we make this a throw error so our tests will fail with a clear reason when we try and bump Clarity if they change the API?

Reply:
I agree

Reply:
As for the concerns, that is how it is being done right now in the VCD ui application for quite some time now and no concerns reported yet